server/src/main/java/org/elasticsearch/search/aggregations/TopBucketBuilder.java

```diff
@@ -20,26 +21,47 @@
 /**
  * Merges many buckets into the "top" buckets as sorted by {@link BucketOrder}.
  */
-public class TopBucketBuilder<B extends InternalMultiBucketAggregation.InternalBucket> {
-    private final Consumer<DelayedBucket<B>> nonCompetitive;
-    private final PriorityQueue<DelayedBucket<B>> queue;
+public abstract class TopBucketBuilder<B extends InternalMultiBucketAggregation.InternalBucket> {
+    /**
+     * The number of buckets required before we switch to the
+     * {@link BufferingTopBucketBuilder}. If we need fewer buckets we use
+     * {@link PriorityQueueTopBucketBuilder}.
+     */
+    static final int USE_BUFFERING_BUILDER = 1024;
 
     /**
      * Create a {@link TopBucketBuilder} to build a list of the top buckets.
+     * <p>
+     * If there are few required results we use a {@link PriorityQueueTopBucketBuilder}
+     * which is simpler and when the priority queue is full but allocates {@code size + 1}
+     * slots in an array. If there are many required results we prefer a
+     * {@link BufferingTopBucketBuilder} which doesn't preallocate and is faster for the
+     * first {@code size} results. But it's a little slower when the priority queue is full.
+     * <p>
+     * It's important for this <strong>not</strong> to preallocate a bunch of memory when
+     * {@code size} is very very large because this backs the reduction of the {@code terms}
+     * aggregation and folks often set the {@code size} of that to something quite large.
+     * The choice in the paragraph above handles this case.
+     *
      * @param size the requested size of the list
      * @param order the sort order of the buckets
      * @param nonCompetitive called with non-competitive buckets
      */
-    public TopBucketBuilder(int size, BucketOrder order, Consumer<DelayedBucket<B>> nonCompetitive) {
-        this.nonCompetitive = nonCompetitive;
-        queue = new PriorityQueue<DelayedBucket<B>>(size) {
-            private final Comparator<DelayedBucket<? extends Bucket>> comparator = order.delayedBucketComparator();
+    public static <B extends InternalMultiBucketAggregation.InternalBucket> TopBucketBuilder<B> build(
+        int size,
+        BucketOrder order,
+        Consumer<DelayedBucket<B>> nonCompetitive
+    ) {
+        if (size < USE_BUFFERING_BUILDER) {
+            return new PriorityQueueTopBucketBuilder<>(size, order, nonCompetitive);
+        }
+        return new BufferingTopBucketBuilder<>(size, order, nonCompetitive);
```

Comment:
I could have made a single TopBucketBuilder which has a buffer and a queue member and flips between them but I didn't feel good making folks that set the size to something low and sensible pay the buffer tax.